### PR TITLE
Fix command handler catchup

### PR DESCRIPTION
**Pull request by vogel**
# Description

This PR makes a few changes that fixes how manual catchup works:

1. it renames manual catchup to offline catchup, and normal one to online catchup
2. in offline catchup buffered ledgers and closing ledger are not supported
3. in offline catchup after applying transaction node goes into LM_SYNCED_STATE (helps with testing)
4. catchup mode is stored in CatchupConfiguration
5. catchup and history tests are simplified, redundant asserts are removed and catchup tests now always wait for LM_SYNCED_STATE state (instead of WAITING_FOR_CLOSING_LEDGER)

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [x] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

**Comment by vogel** [comment]
Looks like I've missed checking computeCatchupPerformedWork in new code, I'll fix that tomorrow.

**Comment by MonsieurNicolas** [comment]
needs to be rebased @vogel

**Comment by vogel** [comment]
rebased

File: src/history/test/HistoryTests.cpp [hunk 1]
```diff
@@ -818,22 +743,12 @@ TEST_CASE("Catchup recent", "[history][catchuprecent]")
 TEST_CASE("Catchup manual", "[history][catchupmanual]")
```
Comment (marta-lokhova) [inline src/history/test/HistoryTests.cpp]: While we are at it: let's not use `std::rand` as it has limited randomness

File: src/main/ApplicationUtils.cpp [hunk 2]
```diff
@@ -372,27 +372,16 @@ catchup(Application::pointer app, CatchupConfiguration cc,
...
-            switch (app->getLedgerManager().getCatchupState())
-            {
-            case LedgerManager::CatchupState::WAITING_FOR_CLOSING_LEDGER:
-            {
-                done = true;
-                synced = true;
-                break;
-            }
-            case LedgerManager::CatchupState::NONE:
+            if (app->getLedgerManager().getCatchupState() ==
```
[the remark this answers is not in the document]
Comment (vogel) [reply src/main/ApplicationUtils.cpp]: `CatchupState::NONE should never occur` in `LM_CATCHING_UP_STATE`. I'll make that an abort() as it would be a logic error.

The splis between catching up states and non-catching up states has two reasons:

1. Before I had to add every LM_CATCHUP_* to every switch statement, and that had no sense at all.

2. The current relationship between `LedgerManager` and `CatchupManager` is broken, and should be fixed (that will require adding a third class that is yet-to-be-named at some point). What I mean is that `LedgerManager` should only operate on catched-up database and do not care about catching up code at all. CatchupState enum should be used only by `CatchupManager`, which should be asking `LedgerManager` to apply ledger for it. A third class should decide where it should send ledgers received from network (to `LedgerManager` if in sync, to `CatchupManager` otherwise). So those two separate enums are there to force us to fix this relationship at some time.

File: src/history/test/HistoryTestsUtils.h [hunk 3]
```diff
@@ -205,6 +205,22 @@ class CatchupSimulation
     std::vector<SequenceNumber> bobSeqs;
     std::vector<SequenceNumber> carolSeqs;
 
+    void ensurePublishesComplete();
```
[the remark this answers is not in the document]
Comment (vogel) [reply src/history/test/HistoryTestsUtils.h]: For verifying state I would just use 'check' or 'verify' names. For me 'ensure' means: check if something is true and if not, do everything to make sure it is true.

For example, `generateAndPublishHistory` was used to generate `n` history checkpoints. New methods, like `ensureOfflineCatchupPossible` are used to make sure that it will be possible to catchup to given ledger. It means that those may or may not generate new ledgers and/or publish them (it depends on current state of the simulation). And it is less iterative and more declarative it its naming, so it should be easier to understand. i.e.:

```
    catchupSimulation.ensureOnlineCatchupPossible(checkpointLedger);
    REQUIRE(catchupSimulation.catchupOnline(app, checkpointLedger));
```

Is much cleaner than previous:

```
    catchupSimulation.generateAndPublishInitialHistory(3);
    catchupSimulation.catchupNewApplication(
        catchupSimulation.getApp().getLedgerManager().getLastClosedLedgerNum() +
            1,
        std::numeric_limits<uint32_t>::max(), false,
        Config::TESTDB_IN_MEMORY_SQLITE, "s3");
```

Here you need to understand relationship between `getLastClosedLedgerNum` and catchup mechanism (and it is non-essential for this test), and also there is not-so-obvious code that adds extra ledgers inside `generateAndPublishInitialHistory` that is not needed for publish, but is needed for catchup. Naming the method `...CatchupPossible` makes it clear, that without those extra ledgers catchup would not be possible.

Also I'm using two functions - `ensureLedgerAvailable` and `ensurePublishesComplete` because it makes them doing only one thing and it allows for better naming. Those are always used together, but that does not mean that using one entity to do two things would be good.

File: src/history/test/HistoryTestsUtils.cpp [hunk 4]
```diff
@@ -489,74 +484,58 @@ CatchupSimulation::generateRandomLedger()
...
-            ((publishSuccesses + nPublishes) * hm.getCheckpointFrequency()) +
-                1);
 }
 
-Application::pointer
-CatchupSimulation::catchupNewApplication(uint32_t initLedger, uint32_t count,
-                                         bool manual, Config::TestDbMode dbMode,
-                                         std::string const& appName)
+void
+CatchupSimulation::ensureOfflineCatchupPossible(uint32_t targetLedger)
```
[the remark this answers is not in the document]
Comment (vogel) [reply src/history/test/HistoryTestsUtils.cpp]: Yes, but that would make tests harder to understand. Instead of asking 'why do I need to make ledgers XYZ available' you can now look at test and see, 'oh, so from now it should be possible to make online/offline catchup to the ledger I've just passed to that method as a parameter'.

File: src/history/test/HistoryTestsUtils.cpp [hunk 5]
```diff
@@ -387,23 +387,18 @@ CatchupSimulation::CatchupSimulation(std::shared_ptr<HistoryConfigurator> cg)
     , mApp(*mAppPtr)
 {
     CHECK(mApp.getHistoryArchiveManager().initializeHistoryArchive("test"));
+    mApp.start();
```
[the remark this answers is not in the document]
Comment (vogel) [reply src/history/test/HistoryTestsUtils.cpp]: This mApp is configured few lines above:

```
mAppPtr(createTestApplication(
          mClock, mHistoryConfigurator->configure(mCfg, true)))
```

so it is OK to start it here

File: src/history/test/HistoryTests.cpp (hunk 1 above)
Comment (vogel) [reply src/history/test/HistoryTests.cpp]: I don't think we need true randomness in that part of the code. Also std::rand is used in some other tests. I think it would be better to replace them with true randomness at once in different PR.

**Comment by vogel** [comment]
Updated.

I've also added few more tests for edge conditions on when catchup is possible.

File: src/history/test/HistoryTestsUtils.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/history/test/HistoryTestsUtils.cpp]: You're right! I was looking at the usage of `createTestApplication` instead

File: src/history/test/HistoryTests.cpp (hunk 1 above)
Comment (marta-lokhova) [reply src/history/test/HistoryTests.cpp]: This can be done in a different PR (happy to look into it when I get a chance), but I disagree that the tests don't need it. Aside from bad randomness, when the default seed is used, rand produces the same sequence every time, and I thought the whole point was that we test different scenarios each time?

File: src/history/test/HistoryTestsUtils.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/history/test/HistoryTestsUtils.h]: I don't feel super strongly about these (personally, just found it a bit hard to follow, and for `ensureLedgerAvailable` and `ensurePublishesComplete` functions I don't really understand what is the test case when we need to use one but not the other? ), so I'm not opposed to keeping it this way

**Comment by marta-lokhova** [comment]
look ok, but you probably need to squash some commits

**Comment by vogel** [comment]
Squashed and rebased.

**Comment by marta-lokhova** [comment]
ah looks like there are conflicts. Maybe it makes sense to rebase again when #1950 is merged (I think it's in the process of being merged right now), then r+ this one.

**Comment by vogel** [comment]
Rebased again

**Comment by marta-lokhova** [comment]
r+ a5ad18a28a67d804d1b3fa4935e89fb0951655a4

**Comment by marta-lokhova** [comment]
Looks like "catchup recent" test failure on gcc

**Comment by MonsieurNicolas** [comment]
needs to be rebased again @vogel

**Comment by vogel** [comment]
rebased

**Comment by marta-lokhova** [comment]
r+ 05a2233ddfff093366f6499b8139291688aeb806

**Comment by MonsieurNicolas** [comment]
failure seemed to be environmental (timeout during the build)

**Comment by MonsieurNicolas** [comment]
@latobarita: retry
